### PR TITLE
AGENT-965: improve node-joiner create command troubleshooting 

### PR DESCRIPTION
**Pull request by andfasano**
This patch introduces a new workflow report object mainly useful for:
* describing in a human-readable format the main workflow events
* capture workflow error

Even though the workflow report could be used for any workflow, currently is enabled just for the add-nodes one.
The workflow report is immediately serialized on disk on each update, so that it could be easily streamed in case of remote execution (`oc adm node-image create` command).

### CODE REVIEW

File: cmd/node-joiner/testdata/add-nodes -report.txt [hunk 1]
```diff
@@ -0,0 +1,24 @@
+
```
Comment (pawanpinjarkar) [inline cmd/node-joiner/testdata/add-nodes -report.txt:1]: The file name seems to have an extra space `cmd/node-joiner/testdata/add-nodes -report.txt` i.e. add-nodes` `-report.txt

File: pkg/asset/agent/workflow/report/context.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	if report, ok := ctx.Value(reportKey).(Report); ok {
+		return report
+	}
+	return nullRepo{}
+}
+
+// nullRepo allows to avoid additional checks when a Report is
+// required, by providing a fully valid (but essentially empty)
+// Report instance.
+type nullRepo struct{}
```
Comment (pawanpinjarkar) [inline pkg/asset/agent/workflow/report/context.go:39]: Just `Report`?

**Review by rwsu — COMMENTED** [review]
/lgtm

File: docs/dev/workflow_report.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+StageFetchBaseISOVerify wr.StageID = wr.NewStageID("fetch-base-iso.verify-version", "Verifying base image version")
+StageFetchBaseISODownload wr.StageID = wr.NewStageID("fetch-base-iso.download-image", "Downloading base ISO image")
+```
+
+### Substages
+Substages are completely optional, and they may be useful for detailing a particularly big or lengthy stage. So, it's perfectly fine to define stages without any substage.
+The current framework does not support more than two levels of stages, essentially for keeping both a simpler interface and for producing an easy to read output for the
+final user.
+
+### (Sub)Stage result
```
Comment (rwsu) [inline docs/dev/workflow_report.md:36]: Thanks for adding lots of examples in this doc. An example of how to add a result may be useful here too.

Comment (rwsu) [reply docs/dev/workflow_report.md:36]: Curious, the current implementation doesn't appear to add any results. Are there more changes to come or is it not currently used by the create command?

Comment (andfasano) [reply docs/dev/workflow_report.md:36]: Yes, the ClusterInfo result will be added in another pr to address the separate card https://issues.redhat.com/browse/AGENT-1000

**Pull request merged**